src/languages/fsharp.js

```diff
@@ -89,52 +78,317 @@ export default function(hljs) {
+    scope: 'string',
+    match: regex.concat(
+      /'/,
+      regex.either(
+          /[^\\']/, // either a single non escaped char...
+          /\\(?:.|\d{3}|x[a-fA-F\d]{2}|u[a-fA-F\d]{4}|U[a-fA-F\d]{8})/ // ...or an escape sequence
+        ),
+      /'/
+    ),
+    relevance: 0
+  };
+  SUBST.contains = [
+    INTERPOLATED_VERBATIM_STRING,
+    INTERPOLATED_STRING,
+    VERBATIM_STRING,
+    FSHARP_CHAR_LITERAL,
+    QUOTED_STRING,
+    hljs.C_NUMBER_MODE,
+    FSHARP_COMMENT
+  ];
+  const FSHARP_STRING = {
+    variants: [
+      INTERPOLATED_TRIPLE_QUOTED_STRING,
+      INTERPOLATED_VERBATIM_STRING,
+      INTERPOLATED_STRING,
+      VERBATIM_STRING,
+      TRIPLE_QUOTED_STRING,
+      QUOTED_STRING,
+      FSHARP_CHAR_LITERAL
+    ]
+  };
+
+  // 'a or ^a
+  const GENERIC_TYPE_SYMBOL = {
+    match: regex.concat(/('|\^)/, hljs.UNDERSCORE_IDENT_RE),
+    scope: 'symbol',
+    relevance: 0
+  };
+
   return {
     name: 'F#',
     aliases: [
       'fs',
       'f#'
     ],
-    keywords: KEYWORDS,
+    keywords: {
+      type: TYPES,
+      keyword: KEYWORDS,
+      literal: LITERALS,
+      built_in: BUILTINS
+    },
     illegal: /\/\*/,
     contains: [
       {
         // monad builder keywords (matches before non-bang keywords)
         scope: 'keyword',
         match: /\b(yield|return|let|do|match|use)!/
       },
+      FSHARP_STRING,
+      FSHARP_COMMENT,
       {
-        scope: 'string',
-        // matches triple quote strings, verbatim strings (@""), character literals...
-        match: /(?:"""[\s\S]*?"""|@"(?:""|[^"])*"|"(?:\\[\s\S]|[^\\"])*")|'(?:[^\\']|\\(?:.|\d{3}|x[a-fA-F\d]{2}|u[a-fA-F\d]{4}|U[a-fA-F\d]{8}))'/
-      },
-      hljs.COMMENT(/\(\*(?!\))/, /\*\)/, {
-        contains: ["self"]
-      }),
-      {
-        // type definitions:
-        scope: 'title.class',
-        beginKeywords: 'type',
-        end: '\\(|=|$',
+        // type MyType<'a> = ...
+        begin: [
+          /type/,
+          /\s+/,
+          hljs.UNDERSCORE_IDENT_RE
+        ],
+        beginScope: {
+          1: 'keyword',
+          3: 'title.class'
+        },
+        end: /\(|=|$/,
         excludeEnd: true,
         contains: [
-          hljs.UNDERSCORE_TITLE_MODE,
-          TYPEPARAM
+          GENERIC_TYPE_SYMBOL
         ]
       },
       {
         // computation expressions:
-        beginScope: "emphasis",
-        match: /\b[_a-z]\w*(?=\s*\{)/i
+        begin: [
+            /\b[_a-z]\w*/,
+            /(?=\s*\{)/
+          ],
+        beginScope: { 1: 'keyword' }
       },
       {
         // preprocessor directives and fsi commands:
-        scope: 'meta',
-        begin: '#',
-        end: '$',
-        keywords: {
-          keyword: 'if else endif line nowarn light r i I load time help quit'
-        }
+        begin: [
+          /^\s*/,
+          regex.concat(/#/, regex.either(...PREPROCESSOR_KEYWORDS)),
+          /\b/
+        ],
+        beginScope: { 2: 'meta' },
+        end: /\s|$/,
```

Comment:
Shouldn't this be a look-ahead?  Remember `\s` also matches newlines. The newline itself shouldn't be part of this match...

Reply:
Ah yet another case where I missed excludeEnd, but the same thougth would apply here: prefer look-ahead. (probably slightly faster also)